text/0143-versioning-and-stability.md

```diff
@@ -0,0 +1,128 @@
+# OpenTelemetry versioning and stability
+
+OpenTelemetry is a large project with strict compatibility requirements. This proposal defines the stability gurantees offered by OpenTelemetry, along with a versioning and lifecycle proposal which defines how we meet those requirements.
+
+Language implementations are expected to follow this proposal exactly, unless a langague or package manager convention interferes significantly. Implementations must take this cross-language proposal, and produce a language-specific proposal which details how these requirements will be met.
+
+## Design goals
+
+**Provide end users with coherent, tested releases.**  
+We do not want users to run untested combinations of core packages. Users must be able to say “I am using OpenTelemetry v1.2.0 for Python” and have a shared understanding as to what that means.
+
+**Ensure that end users stay up to date with the latest release.**  
```

Comment:
Do we have any requirements that it must also be possible for users to only upgrade some, but not all, of OpenTelemetry dependencies? In other words, are languages allowed to have a requirement that versions are aligned? API 1.1 is not guaranteed to work with SDK 1.2, but API 1.2 and SDK 1.2 are guaranteed to work, and support any published instrumentation / public API user out there due to backcompat.

I think in Java it's pretty conventional to require version alignment, and is pretty easy to achieve with BOMs. I think I've seen similar with JS (at least with React), don't know more broadly. Requiring version alignment has the benefit of allowing internal packages to be shared among artifacts within a repository, which I think is useful and sometimes even performance-boosting (e.g., if possible we would only want to introduce a single ThreadLocal string-building buffer in the app, this is only possible if we can share that among all the artifacts as part of an internal package). So I'm hoping that this is an allowed restriction, though can live without it too.

Reply:
What if my application uses several libraries that used Otel API to instrument themselves? These libraries may use different minor versions of API and my app still should work.

Reply:
Yeah - since we "Ensure that end users stay up to date with the latest release.", then it means that this will work no problem if the user uses the latest version of the API or SDK (strictly speaking, it only has to be the version of the latest version specified by a transitive dependency). We know the latest API is compatible with all the libraries (it must be per semver), and we know the lastest API is compatible with the latest SDK. As long as versions are aligned to a new version of things, there's no breakage.

Reply:
@anuraaga I completely agree. I would really prefer that the versions remain aligned, in terms of what libraries are ultimately loaded. Possibly, the API and the SDK need to have a separate version – but I would avoid that if I could. And even in that case, I agree that the user should always load the latest version of the API. Stable packages within the SDK definitely should have the same version number, which matches a tag on the repository. 

We don't want users updating some core packages, but not others. This would result in an untested configuration. I strongly prefer that we only have users running coherent, tested releases. No mix-n-match, please.